### PR TITLE
Fix #7499: Prevent extending java.lang.Enum except from an enum

### LINKED ISSUE #7499: Prevent extending java.lang.Enum except from an enum
as detailed in https://github.com/lampepfl/dotty-feature-requests/issues/73, extending `java.lang.Enum` from a class is forbidden in Java, as Java enums are only guaranteed typesafe when created from its enum construct, additionally, the `super` call to `java/lang/Enum."<init>":(Ljava/lang/String;I)V` is synthesised by `javac` and the constructor is not exposed to Java source programs.

We should do the same in Dotty - extending `java.lang.Enum` should not be permitted apart from in an `enum`.

We should also support the old scala 2 way of extending `java.lang.Enum` with `3.0-migration`

### PR DESCRIPTION
**Pull request by TheElectronWill**
With this PR, extending `java.lang.Enum` from a class or trait gives an error.
```scala
class C1 extends jl.Enum[C1]
      ^
class C1 cannot extend java.lang.Enum: only enums defined with the enum syntax can
```
(one of) The Scala2 way(s) of declaring a java-compatible enum is also covered:
```scala
class C2(name: String, ordinal: Int) extends jl.Enum[C2](name, ordinal)
      ^
class C2 cannot extend java.lang.Enum: only enums defined with the enum syntax can
```
To emit a nice error for the above case I made the two-arguments constructor of `Enum` visible (if there's a better way I'll be glad to change that), and I introduced a check in `CompleteJavaEnums` to prevent that constructor from being explicitely called.
```scala
enum E extends java.lang.Enum[E]("name", 0) {
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            the constructor of java.lang.Enum cannot be called explicitly
```

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -99,10 +100,14 @@ object RefChecks {
...
             cls.sourcePos)
       }
-      for (parent <- cinfo.classParents)
+      val parents = cinfo.classParents
+      for (parent <- parents)
         checkSelfConforms(parent.classSymbol.asClass, "illegal inheritance", "parent")
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
+
+      if !cls.is(Enum) && parents.exists(_.classSymbol == defn.JavaEnumClass) then
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:109]: I believe this should be disabled with source mode `3.0-migration`, (the enum feature is not enabled with that mode)

Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:109]: Secondly, I believe It should be allowed for traits, e.g. we can do the following without issue:
```scala
trait T extends java.lang.Enum[T]

enum TEnum extends T {
  case Foo, Bar
}

object Test extends App {
  println(TEnum.Foo)
}
```

The parents will be deeply checked for any class that extends the trait that is not an enum

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 2]
```diff
@@ -600,7 +605,8 @@ class Definitions {
                       paramNames = Nil, paramInfos = Nil))
                 }
             }
-            constr.info = newInfo
+            val argConstr = constr.copy().entered
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:608]: could you also edit the commit message so that it fixes #7174?

Comment (TheElectronWill) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:608]: yup!

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 1 above)
Comment (TheElectronWill) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: I've added a test for that with the example of #7174

**Comment by TheElectronWill** [comment]
It's now possible to have a class that extend an enum under `-source 3.0-migration` :upside_down_face:

**Comment by bishabosha** [comment]
@TheElectronWill sorry, I made a mistake by suggesting Traits should be able to extend `java.lang.Enum`, e.g. in the test case  
```scala
trait T extends java.lang.Enum[T]

enum MyEnum extends T {
  case A, B
}
```
something like EnumSet becomes unusable due to the F bound:
```scala
scala> java.util.EnumSet.allOf(classOf[MyEnum])
1 |java.util.EnumSet.allOf(classOf[MyEnum])
  |                        ^^^^^^^^^^^^^^^
  |        Found:    (classOf[MyEnum] : Class[MyEnum])
  |        Required: Class[E]
  |
  |        where:    E is a type variable with constraint <: Enum[LazyRef(E)]
```

**Comment by TheElectronWill** [comment]
Ah! It's easy to forbid traits again, I'll make a PR right now.

**Pull request merged**